### PR TITLE
GAWB-849: ga4gh api for agora.

### PR DESCRIPTION
**Pull request by davidangb**
I'm pushing this PR out now for initial review, because it's already gotten large and unwieldy. I'd like to get this preliminary PR merged, then do follow-on work to clean it up. Because this is almost all additive and hidden, I think this is safe.

Must do before merging: a corresponding orchestration PR to reflect that the single ga4gh endpoint exposed by orch has moved, from /ga4gh/ to /ga4gh/v1/. I'll update here once I have that PR.

What this PR does:
* adds ga4gh tool-registry spec (https://github.com/ga4gh/tool-registry-schemas) endpoints for everything that agora can support
* populates the ga4gh model with the minimum necessary data to claim a working endpoint
* copy-and-paste of swagger from the ga4gh spec into agora
* unit tests

What is left to do:
* update swagger to reflect what we _actually_ support, not what the spec supports
* orch passthroughs for all endpoints
* populate the models more fully
* additional unit tests, as appropriate

This google doc: https://docs.google.com/a/broadinstitute.org/spreadsheets/d/1QkgDMlVobQC9t5PjNqMg1nv80IusOm_ovfW7HHmCeGU/edit?usp=sharing
has a summary of progress and level of compatibility for endpoints and models.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [x] **LR** sign off
- [x] **Product Owner** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/resources/swagger/agora.yaml [hunk 1]
```diff
@@ -1322,7 +1322,133 @@ paths:
           schema:
             type: object
             $ref: '#/definitions/Error'
-  /ga4gh/tools/{id}/versions/{version-id}/{type}/descriptor:
+
```
Comment (davidangb) [inline src/main/resources/swagger/agora.yaml:1325]: please read PR description - I know some of this swagger is incorrect and needs updating.

File: src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Ga4ghService.scala [hunk 2]
```diff
@@ -3,36 +3,79 @@ package org.broadinstitute.dsde.agora.server.ga4gh
...
+        path("tools" / Segment / "versions" / Segment / Segment / "descriptor") { (id, versionId, descriptorType) => requestContext =>
+          val entity = entityFromArguments(id, versionId)
+          val message = QueryPublicSinglePayload(requestContext, entity, parseDescriptorType(descriptorType))
+          perRequest(requestContext, queryHandler, message)
+        } ~
+        path("tools" / Segment / "versions" / Segment / Segment / "descriptor" / Segment) { (id, versionId, descriptorType, relativePath) =>
+          complete(spray.http.StatusCodes.NotImplemented)
+        } ~
+        path("tools" / Segment / "versions" / Segment / Segment / "tests") { (id, versionId, descriptorType) =>
+          complete(spray.http.StatusCodes.NotImplemented)
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Ga4ghService.scala:77]: I realize these routes could have been written more hierarchically, e.g. 
```
pathPrefix("tools") {
  pathEnd {} ~
  pathPrefix(Segment) {
    pathEnd {} ~
    pathPrefix("versions") {
      ...
    }
  }
}
```
but after going down that route, I rewrote it "flat" as you can see above; I thought it was more readable.

File: src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Models.scala [hunk 3]
```diff
@@ -13,6 +112,20 @@ object Models {
     val PLAIN_CWL = Value("plain-CWL")
   }
 
+  // ToolTests
+
+  case class ToolDockerfile(
+    dockerfile: String,
+    url: String)
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Models.scala:119]: will delete this model if I can't get its endpoint working in the future

File: src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Models.scala [hunk 4]
```diff
@@ -1,10 +1,109 @@
...
+        description="", // TODO: should we use a snapshot synopsis or description?
+        author="", // TODO: can we get this from the WDL?
+        `meta-version`="", // TODO
+        contains=List.empty[String],
+        verified=false,
+        `verified-source`="",
+        signed=false,
+        versions=entities.toList map (x => ToolVersion(x))
+      )
+    }
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Models.scala:70]: there's a decent bit of business logic in these `apply` methods. It might make more sense to have e.g.:
```
object Tool {
  def apply(entities:Seq[AgoraEntity]) = {
    toolFromEntities(entities)
  }
}
```
and then break out the `toolFromEntities` into some other class. Thoughts?

File: src/main/scala/org/broadinstitute/dsde/agora/server/webservice/handlers/QueryHandler.scala [hunk 5]
```diff
@@ -98,25 +92,6 @@ class QueryHandler(dataSource: PermissionsDataSource, implicit val ec: Execution
...
-          val result = ToolDescriptor(foundEntity.url.getOrElse(""),
-            foundEntity.payload.getOrElse(""),
-            ToolDescriptorType.WDL)
-          RequestComplete(result)
-        case ToolDescriptorType.PLAIN_WDL =>
-          RequestComplete(foundEntity.payload.getOrElse(""))
-      }
-    }
-  }
-
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/agora/server/webservice/handlers/QueryHandler.scala:119]: I moved all the ga4gh-specific handlers into their own dedicated class.

File: src/main/scala/org/broadinstitute/dsde/agora/server/webservice/util/ServiceMessages.scala [hunk 6]
```diff
@@ -76,10 +76,6 @@ object ServiceMessages {
                          onlyPayload: Boolean,
                          payloadAsObject: Boolean)
 
-  case class QueryPublicSingle(requestContext: RequestContext,
-                         entity: AgoraEntity,
-                         descriptorType: DescriptorType)
-
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/agora/server/webservice/util/ServiceMessages.scala:82]: I moved all the ga4gh-specific messages into their own dedicated class.

File: src/test/scala/org/broadinstitute/dsde/agora/server/AgoraTestSuite.scala [hunk 7]
```diff
@@ -28,7 +28,9 @@ class AgoraTestSuite extends Suites(
   new PermissionIntegrationSpec,
   new MethodDefinitionIntegrationSpec,
   new AssociatedConfigurationIntegrationSpec,
+  new CompatibleConfigurationIntegrationSpec,
```
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/agora/server/AgoraTestSuite.scala:31]: this has nothing to do with the current PR, but, uh, we were not running those tests at all.

File: src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Ga4ghService.scala (hunk 2 above)
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Ga4ghService.scala:77]: 👍  for legibility.

File: src/main/resources/swagger/agora.yaml (hunk 1 above)
Comment (rushtong) [reply src/main/resources/swagger/agora.yaml:1325]: This didn't validate for me - I needed to add a `parameters` and `ToolTests` section (which directly mimics the GA4GH spec). I have a PR to this PR with that change: https://github.com/broadinstitute/agora/pull/219

File: src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Ga4ghService.scala (hunk 2 above)
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Ga4ghService.scala:77]: Need to know if we should `TODO` the unimplemented endpoints, or even if they are implementable with what we have.

File: src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Models.scala (hunk 4 above)
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Models.scala:70]: 👍  I like the idea of some kind of `ModelSupport` class that does the heavy lifting and leaving the case classes in their own little world.

File: src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Ga4ghService.scala [hunk 8]
```diff
@@ -3,36 +3,79 @@ package org.broadinstitute.dsde.agora.server.ga4gh
...
-            val toolId = parseId(id)
-
-            val entity = AgoraEntity(Some(toolId.namespace), Some(toolId.name), Some(snapshotId), entityType = Some(AgoraEntityType.Workflow))
-
-            val message = QueryPublicSingle(requestContext, entity, dType)
-            perRequest(requestContext, queryHandler, message)
+    pathPrefix("ga4gh" / "v1") {
+      get {
+        path("metadata") {
+          // because of the dashes in these names, using the automatic spray json marshalling is annoying.
```
Comment (rushtong) [inline src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Ga4ghService.scala:26]: Something to consider is using a different json marshaller, if possible, something like circe which I think the green team is using. Definitely scope creep, but if we run into more difficulties with spray-json, we should consider it.

File: src/test/scala/org/broadinstitute/dsde/agora/server/model/Ga4ghModelTest.scala [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+          `descriptor-type` = List("WDL"),
+          dockerfile = false,
+          `meta-version` = "3",
+          verified = false,
+          `verified-source` = "")
+        val actual = ToolVersion(defaultEntity)
+      }
+    }
+    // TODO: 
+    "ToolDescriptor" - {}
```
Comment (rushtong) [inline src/test/scala/org/broadinstitute/dsde/agora/server/model/Ga4ghModelTest.scala:149]: `    "ToolTests" - {}`

File: src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Models.scala [hunk 10]
```diff
@@ -13,6 +112,20 @@ object Models {
     val PLAIN_CWL = Value("plain-CWL")
   }
 
+  // ToolTests
```
Comment (rushtong) [inline src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Models.scala:115]: Need a TODO for this.

File: src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Models.scala [hunk 11]
```diff
@@ -1,10 +1,109 @@
...
+    def apply(entity:AgoraEntity) = fromEntityType(entity.entityType)
+    def apply(method:MethodDefinition) = fromEntityType(method.entityType)
+    def fromEntityType(entityType: Option[AgoraEntityType.EntityType]) = {
+      assert(entityType.nonEmpty, "cannot create a ToolClass if entity type is empty")
+      val str = entityType.get.toString
+      new ToolClass(str, str, "")
+    }
+  }
+
+  case class Tool(
```
Comment (rushtong) [inline src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Models.scala:37]: Some of these fields are not defined in the GA4GH model, but I'm guessing that's OK? My guess is that as long as we support the minimum, anything extra is gravy. This is also where we'll need alignment with the swagger spec.

**Review by rushtong — COMMENTED** [review]
👍  Looks good. The endpoints worked as expected. Several minor comments inline, but mostly just clarifications. I have one PR that I'll submit against this for the swagger yaml so it validates.

File: src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Ga4ghService.scala (hunk 8 above)
Comment (davidangb) [reply src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Ga4ghService.scala:26]: agreed. I also figured out how to do this with the spray-json, just by using backticks in the model:
```
  case class Metadata(
    version: String,
    `api-version`: String,
    country: String,
    `friendly-name`: String)
```
but I never updated the comment or the endpoint.

File: src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Models.scala (hunk 11 above)
Comment (davidangb) [reply src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Models.scala:37]: which ones are not defined in the GA4GH model? I tried to be exact: https://github.com/ga4gh/tool-registry-schemas/blob/develop/src/main/resources/swagger/ga4gh-tool-discovery.yaml#L335

File: src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Ga4ghService.scala (hunk 2 above)
Comment (davidangb) [reply src/main/scala/org/broadinstitute/dsde/agora/server/ga4gh/Ga4ghService.scala:77]: I don't think we are able to implement the unimplemented endpoints - Agora just doesn't support those things. The relative-path descriptor and tests endpoints are definite no-gos. We could *possibly* implement the dockerfile endpoint, but that would require parsing the WDL and pulling out the info ... but even then, within a Workflow, each Task might have a different dockerfile, and so which should we report?

So, my plan was:
* remove swagger defs for these endpoints, so we only show swagger for what we actually support
* remove models, tests, TODOs, and any code that references these endpoints
* BUT, leave the route in place and respond with NotImplemented (instead of NotFound) just in case a power user attempts the route - this is a signal that we know the route should exist per the spec, but we don't have anything there.
